File: paperjs/_scripts/mesh/PointFollower.js
/*jslint vars: true, nomen: true, plusplus: true, continue:true, forin:true */
/*global atob, btoa, ArrayBuffer, Uint8Array, Blob */

class PointFollower extends Follower {

	//var target = null;
	//var attractionCoefficient = 0.1;

	constructor (points = []) {
		super();
		this.points = points;
		this.attractionCoefficient = 0.1;
		this.pointIndex = 0;
		this.hitRadius = 10;
		this.randomOrder = false;
		this.pathJitter = 5;
		this.currentTarget = null;
		this.loops = 0;
	}


	getCurrentTarget () {
		//return this.points[this.pointIndex];

		if(!this.currentTarget) {
			this.currentTarget = this.points[0];
		}

		return this.currentTarget;
	}

	getNextPoint (){

		var p;
		if(this.randomOrder) {
			var index = Math.floor(Math.random() * this.points.length);
			this.pointIndex = index;
			p = this.points[index];
		} else {
			this.pointIndex++;
			if(this.pointIndex == this.points.length) {

				this.loops++;
				this.pointIndex = 0;
			}

			p = this.points[this.pointIndex];
		}

		if(this.pathJitter) {
			p = Utils.randomPointOnCircle(p, Math.random() * this.pathJitter);
		}

		return p;
	}

	update() {

		if(!this.points.length) {
			return;
		}

		var tPoint = this.getCurrentTarget();

		if(tPoint.getDistance(this.location) < this.hitRadius) {
			tPoint = this.getNextPoint();

			this.currentTarget = tPoint;

			if(this.onPointHit) {
				//this.onPointHit(oldPoint, newPoint);
			}
		}

		super.update(tPoint);
		//can call super right here

    	//todo
    	/*
    	this.velocity = this.velocity.normalize(this.limit);

    	var dir = tPoint.subtract(this.location);

    	dir = dir.normalize();

    	//+=
    	dir = dir.multiply(this.attractionCoefficient);

	    this.acceleration = this.acceleration.add(dir);
	    this.velocity = this.velocity.add(this.acceleration);
	    this.location = this.location.add(this.velocity);
	    
	    this.updateAngle();
	    this.acceleration.set(0,0);
	    */
    }
}

File: openframeworks/lib/mesh/PointFollower.cpp
#include "PointFollower.h"
#include "MeshUtils.h"

PointFollower::PointFollower() {
    attractionCoefficient = 1.8;
    
    currentPoint = ofVec3f(0, 0, CURRENT_POINT_INITIALIZATION_VALUE);
}

void PointFollower::setPoints(vector<ofVec3f> _points) {
    points = _points;
    pointIndex = 0;
}

ofVec3f PointFollower::getCurrentPoint () {
    
    if(pointIndex == -1) {
        cout << "PointFollower::getCurrentTarget points have not been specified" << endl;
        return;
    }
    
    //if it hasnt been initialized yet, get the first point
    if(currentPoint.z == CURRENT_POINT_INITIALIZATION_VALUE) {
        currentPoint = points.at(0);
        pointIndex = 0;
    }
    
    //note: need to check if currentPoint has been initialized
    return currentPoint;
    
}

//should probably make this private
ofVec3f PointFollower::getNextPoint (){
    
    ofVec3f p;
    if(randomOrder) {
        int index = floor(ofRandom(1) * points.size());
        pointIndex = index;
        p = points.at(index);
    } else {
        pointIndex++;
        if(pointIndex == points.size()) {
            loops++;
            pointIndex = 0;
        }
        
        p = points.at(pointIndex);
    }
    
    if(pathJitter) {
        float a = ofRandom(0, pathJitter);
        p = mGetRandomPointOnCircle(p, a);
    }
    
    return p;
}

void PointFollower::update() {
    
    if(pointIndex == -1) {
        return;
    }
    
    ofVec3f tPoint = getCurrentPoint();
    
    if(ofDist(tPoint.x, tPoint.y, location.x, location.y) < hitRadius) {
        tPoint = getNextPoint();
    
        currentPoint = tPoint;
    }

    Follower::update(tPoint);
}
